File: LaboratoryWork/Semester_3/Laboratory_2/GoLang/tasks/arithmeticExpression.go
package tasks

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"unicode"
)

// Stack — простая реализация стека
type Stack[T any] struct {
	data []T
}

func (s *Stack[T]) Push(val T) {
	s.data = append(s.data, val)
}

func (s *Stack[T]) Pop() (T, error) {
	if len(s.data) == 0 {
		var zero T
		return zero, errors.New("stack is empty")
	}
	val := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return val, nil
}

func (s *Stack[T]) Top() (T, error) {
	if len(s.data) == 0 {
		var zero T
		return zero, errors.New("stack is empty")
	}
	return s.data[len(s.data)-1], nil
}

func (s *Stack[T]) Empty() bool {
	return len(s.data) == 0
}

// --- Вспомогательные функции ---

func isDigit(r rune) bool {
	return unicode.IsDigit(r)
}

func isOpeningBracket(r rune) bool {
	return r == '('
}

func isClosingBracket(r rune) bool {
	return r == ')'
}

func getPriority(op rune) int {
	switch op {
	case '+', '-':
		return 1
	case '*':
		return 2
	default:
		return 0
	}
}

func calculate(a, b int64, op rune) int64 {
	switch op {
	case '+':
		return a + b
	case '-':
		return a - b
	case '*':
		return a * b
	default:
		return 0
	}
}

// --- Основная функция вычисления выражения ---

func ArithmeticExpression() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Введите свое арифметическое выражение: ")
	expr, _ := reader.ReadString('\n')

	var operators Stack[rune]
	var operands Stack[int64]

	runes := []rune(expr)

	for i := 0; i < len(runes); {
		ch := runes[i]

		if unicode.IsSpace(ch) {
			i++
			continue
		}

		if isDigit(ch) {
			var num int64
			for i < len(runes) && isDigit(runes[i]) {
				num = num*10 + int64(runes[i]-'0')
				i++
			}
			operands.Push(num)
			continue
		}

		if isOpeningBracket(ch) {
			operators.Push(ch)
			i++
			continue
		}

		if isClosingBracket(ch) {
			for !operators.Empty() {
				top, _ := operators.Top()
				if top == '(' {
					break
				}
				operators.Pop()

				b, err2 := operands.Pop()
				a, err1 := operands.Pop()
				if err1 != nil || err2 != nil {
					fmt.Println("[ ERROR ] Missing operand!")
					return
				}

				res := calculate(a, b, top)
				operands.Push(res)
			}

			if operators.Empty() {
				fmt.Println("[ ERROR ] Opening bracket is missing!")
				return
			}
			operators.Pop() // удалить '('
			i++
			continue
		}

		if ch == '+' || ch == '-' || ch == '*' {
			for !operators.Empty() {
				top, _ := operators.Top()
				if getPriority(top) < getPriority(ch) {
					break
				}
				operators.Pop()

				b, err2 := operands.Pop()
				a, err1 := operands.Pop()
				if err1 != nil || err2 != nil {
					fmt.Println("[ ERROR ] Missing operand!")
					return
				}

				operands.Push(calculate(a, b, top))
			}
			operators.Push(ch)
			i++
			continue
		}

		fmt.Println("[ ERROR ] Invalid symbol in expression!")
		return
	}

	for !operators.Empty() {
		op, _ := operators.Pop()
		b, err2 := operands.Pop()
		a, err1 := operands.Pop()
		if err1 != nil || err2 != nil {
			fmt.Println("[ ERROR ] Missing operand!")
			return
		}
		operands.Push(calculate(a, b, op))
	}

	if operands.Empty() {
		fmt.Println("[ ERROR ] Expression is empty!")
		return
	}

	res, _ := operands.Top()
	fmt.Printf("Значение выражения: %d\n", res)
}

File: LaboratoryWork/Semester_3/Laboratory_2/C++/src/tasks/arithmetic_expression.cpp
#include <iostream>
#include <string>
#include <cstdint>
#include "../../include/stack.hpp"

using namespace std;

bool isSpace(char c) {
    return c == ' ';
}

bool isDigit(char c) {
    return (c >= '0' && c <= '9');
}

bool isOpeningBracket(char c) {
    return c == '(';
}

bool isClosingBracket(char c) {
    return c == ')';
}

int64_t calculate(int64_t firstOperand, int64_t secondOperand, char op) {
    switch (op) {
        case '+':
            return firstOperand + secondOperand;
        case '-':
            return firstOperand - secondOperand;
        case '*':
            return firstOperand * secondOperand;
    }
    return 0;
}

int64_t getPriority(char op) {
    if (op == '+' || op == '-') {
        return 1;
    } else if (op == '*') {
        return 2;
    }
    return 0;
}

void arithmeticExpression() {
    cout << "Введите свое арифметическое выражение: ";

    string expression;
    getline(cin, expression);

    Stack<char> operators;
    Stack<int64_t> operands;

    for (int i = 0; i < expression.length();) {
        char currentChar = expression[i];
        
        if (isSpace(currentChar)) {
            i++;
            continue;
        }

        if (isDigit(currentChar)) { 
            int64_t currentNum = 0;
            while (i < expression.length() && isDigit(expression[i])) {
                currentNum = currentNum * 10 + (expression[i] - '0');
                i++;
            }
            operands.Push(currentNum);
            continue;
        }

        if (isOpeningBracket(currentChar)) {
            operators.Push(currentChar);
            i++;
            continue;
        }

        if (isClosingBracket(currentChar)) {
            while (!operators.Empty() && operators.Top() != '(') {
                if (operands.Empty()) {
                    throw runtime_error("[ ERROR ] Missing operand!");
                }

                int64_t secondOperand = operands.Top();
                operands.Pop();
                
                if (operands.Empty()) {
                    throw runtime_error("[ ERROR ] Missing operand!");
                }
                
                int64_t firstOperand = operands.Top();
                operands.Pop();

                char currentOperator = operators.Top();
                operators.Pop();

                operands.Push(calculate(firstOperand, secondOperand, currentOperator));
            }

            if (operators.Empty()) {
                throw runtime_error("[ ERROR ] Opening bracket is missing!");
            }

            operators.Pop();
            i++;
            continue;
        }

        if (currentChar == '+' || currentChar == '-' || currentChar == '*') {
            while (!operators.Empty() && getPriority(operators.Top()) >= getPriority(currentChar)) {
                char op = operators.Top();
                operators.Pop();

                if (operands.Empty()) {
                    throw runtime_error("[ ERROR ] Missing operand!");
                }

                int64_t secondOperand = operands.Top();
                operands.Pop();

                if (operands.Empty()) {
                    throw runtime_error("[ ERROR ] Missing operand!");
                }

                int64_t firstOperand = operands.Top();
                operands.Pop();

                operands.Push(calculate(firstOperand, secondOperand, op));
            }
            operators.Push(currentChar);
            i++;
            continue;
        }

        throw runtime_error("[ ERROR ] Invalid symbol in expression!");
    }

    while (!operators.Empty()) {
        char op = operators.Top();
        operators.Pop();

        if (operands.Empty()) {
            throw runtime_error("[ ERROR ] Missing operand!");
        }

        int64_t secondOperand = operands.Top();
        operands.Pop();

        if (operands.Empty()) {
            throw runtime_error("[ ERROR ] Missing operand!");
        }

        int64_t firstOperand = operands.Top();
        operands.Pop();

        operands.Push(calculate(firstOperand, secondOperand, op));
    }

    if (operands.Empty()) {
        throw runtime_error("[ ERROR ] Expression is empty!");
    }

    cout << "Значение выражения: " << operands.Top() << endl;
}